### PR TITLE
Allow a Database to be forked off of a Mirror

### PR DESCRIPTION
**Pull request by mxgrey**
### Implemented feature

Now it is possible for a schedule Mirror to fork off a new schedule Database whose state will match the version of the upstream Database that the Mirror is synced with.

### CODE REVIEW

File: rmf_traffic/test/unit/schedule/test_Mirror.cpp [hunk 1]
```diff
@@ -487,3 +488,304 @@ SCENARIO("Testing specialized mirrors")
...
+  rmf_traffic::schedule::Mirror _mirror;
+  std::shared_ptr<rmf_traffic::schedule::DatabaseRectificationRequesterFactory>
+  _rectifiers;
+
+  rmf_traffic::schedule::Query _all = rmf_traffic::schedule::query_all();
+
+};
+
+//==============================================================================
+SCENARIO("Test forking off of mirrors")
```
Comment (gbiggs) [inline rmf_traffic/test/unit/schedule/test_Mirror.cpp:639]: I know it's only tests, but these could do with a few comments to break up and explain the specific things being tested.

**Review by gbiggs — APPROVED** [review]

File: rmf_traffic/include/rmf_traffic/schedule/Mirror.hpp [hunk 2]
```diff
@@ -83,6 +83,11 @@ class Mirror : public ItineraryViewer, public Snappable
   /// patch does not match
   bool update(const Patch& patch);
 
+  /// Fork a new database off of this Mirror. The state of the new database
+  /// will match the state that the upstream Database had at the time it sent
```
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic/include/rmf_traffic/schedule/Mirror.hpp:87]: That's a good point. I'll tweak the wording here.

File: rmf_traffic/test/unit/schedule/test_Mirror.cpp (hunk 1 above)
Comment (mxgrey) [reply rmf_traffic/test/unit/schedule/test_Mirror.cpp:639]: Honestly I'm just have it do a bunch of arbitrary operations on the participant, some of which get dropped and then rectified later, all while the database fails over several times, and I just want to see that after each rectification the Database information always matches the participant information, no matter how or when the failovers occur.

I guess I can at least add an explanation to that effect in the test.

File: rmf_traffic/include/rmf_traffic/schedule/Mirror.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mxgrey) [reply rmf_traffic/include/rmf_traffic/schedule/Mirror.hpp]: Fixed: 93a96e3

File: rmf_traffic/test/unit/schedule/test_Mirror.cpp (hunk 1 above)
Comment (mxgrey) [reply rmf_traffic/test/unit/schedule/test_Mirror.cpp:639]: Updated: 93a96e3

**Review by gbiggs — APPROVED** [review]

**Pull request merged**